packages/server-renderer/src/render.ts

```diff
@@ -94,21 +94,20 @@ export function renderComponentVNode(
   const instance = createComponentInstance(vnode, parentComponent, null)
   const res = setupComponent(instance, true /* isSSR */)
   const hasAsyncSetup = isPromise(res)
-  const prefetches = instance.sp /* LifecycleHooks.SERVER_PREFETCH */
+  let prefetches = instance.sp /* LifecycleHooks.SERVER_PREFETCH */
   if (hasAsyncSetup || prefetches) {
-    let p: Promise<unknown> = hasAsyncSetup
-      ? (res as Promise<void>)
-      : Promise.resolve()
-    if (prefetches) {
-      p = p
-        .then(() =>
-          Promise.all(
+    const p: Promise<unknown> = Promise.resolve(res as Promise<void>)
+      .then(() => {
+        // instance.sp may be null until an async setup resolves, so evaluate it here
+        prefetches = instance.sp
```

Comment:
```suggestion
        if(hasAsyncSetup) prefetches = instance.sp
```

Reply:
I still believe here should check `hasAsyncSetup`

Reply:
Please see commit https://github.com/vuejs/core/pull/10893/commits/7bf6732e17e4f6cee6c0bc3a149fbc5a9258c404